### PR TITLE
Tusk

### PR DESCRIPTION
**Pull request by Wouterz90**

### CODE REVIEW

File: game/scripts/npc/npc_heroes_custom.txt [hunk 1]
```diff
@@ -4,6 +4,27 @@
 
 "DOTAHeroes"
 {
+    // Vengeful Spirit correct as at 7.07c
```
Comment (Perryvw) [inline game/scripts/npc/npc_heroes_custom.txt]: I dont think this is actually English

File: game/scripts/vscripts/heroes/tusk/snowball.lua [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    --[MODIFIER_STATE_INVULNERABLE] = true, -- These state make allied orders impossible
+    --[MODIFIER_STATE_OUT_OF_GAME] = true,
+    [MODIFIER_STATE_NO_HEALTH_BAR] = true,
+    [MODIFIER_STATE_UNSELECTABLE] = false,
+  }
+end
+
+---@return
+---@param keys table
+  function modifier_tusk_snowball_host:OnOrder(keys)
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/snowball.lua]: This should technically be an override although I guess that breaks with the current API docs.

File: game/scripts/vscripts/heroes/tusk/snowball.lua [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+
+---@override
+function modifier_tusk_snowball_guest:OnDestroy()
+  if IsServer() then -- Nothing needs to be done on the client
+    self:GetParent():RemoveNoDraw()
+  end
+end
+
+---@return table
+function modifier_tusk_snowball_guest :CheckState()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/snowball.lua]: Also technically override

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+end
+-- This does show a damage block icon, but I think it's better than setting health
+---@override
+function modifier_tusk_sigil_slow_aura:GetModifierTotal_ConstantBlock(keys)
+  local damage = keys.damage
+  local attacker = keys.attacker
+  -- Block all damage from abilities
+  if keys.inflictor then return keys.damage end
+  if attacker:IsRealHero() then
+    return keys.damage - 4
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: Comment explaining magic numbers please

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  else
+    return keys.damage - 1
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: Another magic number

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+---@override
+function modifier_tusk_sigil_slow_aura:GetModifierAura()
+  return "modifier_tusk_sigil_slow_aura_modifier"
+end
+---@return DOTA_UNIT_TARGET_TYPE
+function modifier_tusk_sigil_slow_aura:GetAuraSearchType()
+  return DOTA_UNIT_TARGET_HERO + DOTA_UNIT_TARGET_BASIC
+end
+---@return table
+function modifier_tusk_sigil_slow_aura:DeclareFunctions()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: Should sigil not also be magic immune?

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  -- Default modifier, can't be recreated.
+  unit:AddNewModifier(caster,ability,"modifier_kill",{duration = sigil_duration})
+  unit:AddNewModifier(caster,ability,"modifier_tusk_sigil_slow_aura",{duration = sigil_duration})
+end
+
+LinkLuaModifier("modifier_tusk_sigil_slow_aura","heroes/tusk/frozen_sigil.lua",LUA_MODIFIER_MOTION_NONE)
+---@class modifier_tusk_sigil_slow_aura : CDOTA_Modifier_Lua
+modifier_tusk_sigil_slow_aura = class({})
+
+function modifier_tusk_sigil_slow_aura:OnCreated()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: ---@override

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+end
+---@return table
+function modifier_tusk_sigil_slow_aura:DeclareFunctions()
+  return {
+    MODIFIER_PROPERTY_TOTAL_CONSTANT_BLOCK
+  }
+end
+-- This does show a damage block icon, but I think it's better than setting health
+---@override
+function modifier_tusk_sigil_slow_aura:GetModifierTotal_ConstantBlock(keys)
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: Sigil health scales per level 3/3/4/4 hits from heroes, 12/12/16/16 from creeps, where are the different health values?

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+---@override
+function modifier_tusk_sigil_slow_aura:IsDebuff() return false end
+---@override
+function modifier_tusk_sigil_slow_aura:GetAuraSearchTeam()
+  return DOTA_UNIT_TARGET_TEAM_ENEMY
+end
+---@override
+function modifier_tusk_sigil_slow_aura:GetAuraRadius()
+
+  return self:GetAbility():GetSpecialValueFor("sigil_radius")
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: I dont think an existing sigil changes if the ability is leveled, if this is the case I would move this to OnCreated.

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+tusk_frozen_sigil_lua = class({})
+---@override
+function tusk_frozen_sigil_lua:OnSpellStart()
+  local caster = self:GetCaster()
+  local sigil_duration = self:GetSpecialValueFor("sigil_duration")
+  local unit = CreateUnitByName("npc_dota_tusk_frozen_sigil"..self:GetLevel(),caster:GetAbsOrigin(),false,caster,caster:GetPlayerOwner(),caster:GetTeamNumber())
+  unit:SetControllableByPlayer(caster:GetPlayerOwnerID(),false)
+  --Transfer this ability to the sigil, so there can't be nil references when this is stolen
+  local ability = unit:AddAbility(self:GetAbilityName())
+  ability:SetLevel(self:GetLevel())
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: maybe set it hidden on the sigil, check how the game does it.

File: game/scripts/vscripts/heroes/tusk/snowball.lua [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+    if not self.unitsHit[unit] then
+      local damage_table = {
+        victim = unit,
+        attacker = caster,
+        ability = self,
+        damage = damage + (#self.unitInSnowball-1 * damage_per_unit),
+        damage_type = self:GetAbilityDamageType(),
+      }
+      ApplyDamage(damage_table)
+      unit:AddNewModifier(caster,self,"modifier_stunned",{duration = stun_duration})
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/snowball.lua]: Why is this not in OnProjectileHit, would this not apply multiple times to the same units

File: game/scripts/vscripts/heroes/tusk/snowball.lua [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+  if IsServer() then -- Nothing needs to be done on the client
+    -- Store all the values
+    local ability = self:GetAbility()
+
+    self.snowball_grow_rate = ability:GetSpecialValueFor("snowball_grow_rate")
+    self.snowball_grab_radius = ability:GetSpecialValueFor("snowball_grab_radius")
+
+    self.units_inside = 0
+
+    --caster:AddNoDraw()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/snowball.lua]: This is probably right

File: game/scripts/vscripts/heroes/tusk/snowball.lua [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+    --self:GetCaster():RemoveNoDraw()
+  end
+end
+
+---@return table
+function modifier_tusk_snowball_host:DeclareFunctions()
+  return {
+    MODIFIER_EVENT_ON_ORDER,
+    MODIFIER_PROPERTY_INCOMING_DAMAGE_PERCENTAGE,
+    MODIFIER_PROPERTY_MODEL_CHANGE,
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/snowball.lua]: Would ditch the model change and go for nodraw + particle personally

File: game/scripts/vscripts/heroes/tusk/snowball.lua [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+
+    self:GetParent():EmitSound("Hero_Tusk.Snowball.Ally")
+    self:GetParent():AddNoDraw()
+  end
+end
+
+---@override
+function modifier_tusk_snowball_guest:OnDestroy()
+  if IsServer() then -- Nothing needs to be done on the client
+    self:GetParent():RemoveNoDraw()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/snowball.lua]: Maybe add a FindClearSpaceForUnit here

File: game/scripts/vscripts/heroes/tusk/walrus_kick.lua [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  unit:SetAbsOrigin(unit:GetAbsOrigin() + self.direction *  FrameTime())
+end
+
+---@override
+function modifier_tusk_walrus_kick_flying:OnDestroy()
+  if IsServer() then
+    -- Make sure the unit ends on the ground
+    -- Don't think this is needed though, the engine sets unit to ground level on movement
+    local origin = GetGroundPosition(self:GetParent():GetAbsOrigin(),self:GetParent())
+    self:GetParent():SetAbsOrigin(origin)
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/walrus_kick.lua]: I think you can do both of these lines but better using FindClearSpaceForUnit

File: game/scripts/vscripts/heroes/tusk/walrus_punch.lua [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+end
+
+---@override
+function modifier_tusk_walrus_punch_crit:OnAttackLanded()
+  self:GetCaster():EmitSound("Hero_Tusk.WalrusPunch.Target")
+  self:Destroy()
+end
+LinkLuaModifier("modifier_tusk_walrus_punch_flying","heroes/tusk/walrus_punch.lua",LUA_MODIFIER_MOTION_NONE)
+---@class modifier_tusk_walrus_punch_flying : CDOTA_Modifier_Lua
+modifier_tusk_walrus_punch_flying = class({})
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/tusk/walrus_punch.lua]: Is this not just a knockback modifier? I'm not sure we need to recreate this.

**Review by Perryvw — CHANGES_REQUESTED** [review]
Global comments:
Indentation: 4 spaces
File organisation:
```
All LinkLuaModifier calls

All modifiers

Ability code
```

File: game/scripts/vscripts/heroes/tusk/frozen_sigil.lua (hunk 8 above)
Comment (Wouterz90) [reply game/scripts/vscripts/heroes/tusk/frozen_sigil.lua]: There are 4 units. npc_dota_tusk_frozen_sigil1 to npc_dota_tusk_frozen_sigil4. The health is set in those files.

File: game/scripts/vscripts/heroes/tusk/snowball.lua (hunk 11 above)
Comment (Wouterz90) [reply game/scripts/vscripts/heroes/tusk/snowball.lua]: It's a tracking projectile, I would be surprised if it triggers reaching units that are not the target. 
I did indeed forget to self.unitsHit[unit] = true

File: game/scripts/vscripts/heroes/tusk/snowball.lua (hunk 14 above)
Comment (Wouterz90) [reply game/scripts/vscripts/heroes/tusk/snowball.lua]: ResolveNPCPositions is used when the projectile hits

**Review by Perryvw — APPROVED** [review]

**Pull request merged**